### ISSUE TITLE
Tracker detector IDs in Java and LCDD do not match

### ISSUE THREAD

**Issue by JeremyMcCormick** [JeffersonLab/hps-java#108]
The "module" field is not assigned consistently between the LCDD and Java versions of the tracker geometries for `HPS-EngRun2015-Nominal-v5-0`.

From debug print outs added to `SimTrackerHitReadoutDriver` in lcsim:

```
SimTrackerHitReadoutDriver: proc hit
  id: /1/0/7/2/0/0/0
  found 1 detelem
    module_L4t_halfmodule_axial_slot_sensor0 with ID /1/0/7/2/0/0/0
  found detelem 'module_L4t_halfmodule_axial_slot_sensor0' using ID lkp with id = /1/0/7/2/0/0/0
  found detelem 'module_L4t_halfmodule_axial_hole_sensor0' by pos with id = /1/0/7/0/0/0/0
  ERROR: detelems found from pos and id do not match!!!
  used detelem 'module_L4t_halfmodule_axial_hole_sensor0' at [     -35.463,      26.639,      489.65]

SimTrackerHitReadoutDriver: proc hit
  id: /1/0/8/2/0/0/0
  found 1 detelem
    module_L4t_halfmodule_stereo_slot_sensor0 with ID /1/0/8/2/0/0/0
  found detelem 'module_L4t_halfmodule_stereo_slot_sensor0' using ID lkp with id = /1/0/8/2/0/0/0
  found detelem 'module_L4t_halfmodule_stereo_hole_sensor0' by pos with id = /1/0/8/0/0/0/0
  ERROR: detelems found from pos and id do not match!!!
  used detelem 'module_L4t_halfmodule_stereo_hole_sensor0' at [     -35.172,      29.158,      497.12]
```

Checking the LCDD file for one of the detectors shows this mismatch:

```
<physvol>
    <volumeref ref="module_L6t_halfmodule_axial_hole_volume" />
    <positionref ref="module_L6t_halfmodule_axial_hole_position" />
    <rotationref ref="module_L6t_halfmodule_axial_hole_rotation" />
    <physvolid field_name="system" value="1" />
    <physvolid field_name="barrel" value="0" />
    <physvolid field_name="layer" value="11" />
    <physvolid field_name="module" value="2" />
</physvol>
```

Dump of Java detector layers 11 and 12:

```
  Tracker_layer12: /1/0/12/0/0/0/0
      module_L6b_halfmodule_axial_hole: /0/0/0/0/0/0/0
        module_L6b_halfmodule_axial_hole_sensor0: /1/0/12/1/0/0/0
      module_L6b_halfmodule_axial_slot: /0/0/0/0/0/0/0
        module_L6b_halfmodule_axial_slot_sensor0: /1/0/12/3/0/0/0
      module_L6t_halfmodule_stereo_hole: /0/0/0/0/0/0/0
        module_L6t_halfmodule_stereo_hole_sensor0: /1/0/12/0/0/0/0
      module_L6t_halfmodule_stereo_slot: /0/0/0/0/0/0/0
        module_L6t_halfmodule_stereo_slot_sensor0: /1/0/12/2/0/0/0
    Tracker_layer11: /1/0/11/0/0/0/0
      module_L6b_halfmodule_stereo_hole: /0/0/0/0/0/0/0
        module_L6b_halfmodule_stereo_hole_sensor0: /1/0/11/1/0/0/0
      module_L6b_halfmodule_stereo_slot: /0/0/0/0/0/0/0
        module_L6b_halfmodule_stereo_slot_sensor0: /1/0/11/3/0/0/0
      module_L6t_halfmodule_axial_hole: /0/0/0/0/0/0/0
        module_L6t_halfmodule_axial_hole_sensor0: /1/0/11/0/0/0/0
      module_L6t_halfmodule_axial_slot: /0/0/0/0/0/0/0
        module_L6t_halfmodule_axial_slot_sensor0: /1/0/11/2/0/0/0
```

In the Java detector the module for this component is 0 not 2.

Problem seems to be with the results from this method call:

```
int moduleNumber = getModuleNumber(surveyVolume.getName());
```

The method is overloaded to use a different method when outputting LCDD which does not use the same scheme as the corresponding method for the in-memory Java detector.

**Comment by omar-moreno** [comment]
I'm not familar with this code so I wont be much help here. This is
definitely a project you should take on.

**Comment by JeremyMcCormick** [comment]
Do you know how module is being used here?

Also what are the implications of changing the module number value on the
detector components for track recon, if any?

**Comment by normangraf** [comment]
We should determine whether the identifiers are at least unique. If so, we can create a mapping from wrong to right. If the IDs are not unique, for example, if multiple sensors are being mapped to a single ID, then there is no fix.

**Comment by perhansson** [comment]
I can look into it.


Jeremy, can you write a very short instruction on how to get things running with the environment you have so that I can play with it?



________________________________

**Comment by mholtrop** [comment]
There would still be a fix, even with multiple mapped to single ID, you just end up with more complicated disambiguation. In the vain of what we are doing currently, you can use the position to figure out how to map, but you can do so much faster if the options are limited.

**Comment by JeremyMcCormick** [comment]
@perhansson 

To reproduce this, you can run the readout simulation:

```
java -DdisableSvtAlignmentConstants -Dorg.hps.conditions.connection.resource=/org/hps/conditions/config/slac_connection.prop -jar ~/.m2/repository/org/hps/hps-distribution/3.11-SNAPSHOT/hps-distribution-3.11-SNAPSHOT-bin.jar -i tritrig_filt.slcio -D outputFile=tritrig_readout -D detector=HPS-EngRun2015-Nominal-v5-0-fieldmap -D run=5772 -n 5 -r /org/hps/steering/readout/EngineeringRun2015TrigPairs1_Pass2.lcsim
```

Input file is here at SLAC:

```
/nfs/slac/g/hps3/data/test/tritrig_filt.slcio
```

I use a hacked up copy of this lcsim Driver to print out debug statements:

http://www.slac.stanford.edu/~jeremym/hps/SimTrackerHitReadoutDriver.java

To see the LCDD output I've added debug print outs to hps-java and I run:

```
java -jar ./hps-detector-model-3.11-SNAPSHOT-bin.jar -o LCDD compact.xml det.lcdd
```

**Comment by perhansson** [comment]
Ok, I’ll give it a try.




On Jun 6, 2017, at 1:46 PM, Jeremy McCormick <notifications@github.com<mailto:notifications@github.com>> wrote:


@perhansson<https://github.com/perhansson>

To reproduce this, you can run the readout simulation:

java -DdisableSvtAlignmentConstants -Dorg.hps.conditions.connection.resource=/org/hps/conditions/config/slac_connection.prop -jar ~/.m2/repository/org/hps/hps-distribution/3.11-SNAPSHOT/hps-distribution-3.11-SNAPSHOT-bin.jar -i tritrig_filt.slcio -D outputFile=tritrig_readout -D detector=HPS-EngRun2015-Nominal-v5-0-fieldmap -D run=5772 -n 5 -r /org/hps/steering/readout/EngineeringRun2015TrigPairs1_Pass2.lcsim


Input file is here at SLAC:

/nfs/slac/g/hps3/data/test/tritrig_filt.slcio


I use a hacked up copy of this lcsim Driver to print out debug statements:

http://www.slac.stanford.edu/~jeremym/hps/SimTrackerHitReadoutDriver.java<http://www.slac.stanford.edu/%7Ejeremym/hps/SimTrackerHitReadoutDriver.java>

To see the LCDD output I've added debug print outs to hps-java and I run:

java -jar ./hps-detector-model-3.11-SNAPSHOT-bin.jar -o LCDD compact.xml det.lcdd

**Comment by perhansson** [comment]
#110 may fix this. I looked at it and it was just swapped. I don't know why and I don't see any reason for it being swapped which probably means it's just a bug that never got discovered because nobody cared of the LCDD module ID until now.

**Comment by perhansson** [comment]
I haven't fully tested with the java geometry but the LCDD geometry was swapped now.

**Comment by JeremyMcCormick** [comment]
Thanks for the patch @perhansson.

I will take a look at it....

**Comment by omar-moreno** [comment]
Given that this change is going to make all our Monte Carlo useless, we should notify Bradley to stop generating events until after this change is in right?

**Comment by JeremyMcCormick** [comment]
It should be fine to merge in Pelle's fix, actually.  Existing MC files
will still be usable because the ID is not currently used; the hit is
connected to the sensor using the position.  Then the sensor's ID info is
used instead.

What we would want to decide is whether we should remake the affected LCDD
files as well as regenerate existing MC samples so that they are fixed and
the IDs are usable.

To improve the performance of hit assignment in readout I have a Driver
which I'll add to another branch that uses the layer number to reduce the
number of detector elements used for the position lookup.  The LCDD IDs in
layers with 2 modules appear to be fine so I believe those can be used
directly.

**Comment by normangraf** [comment]
Hello Omar,

I don't think it would change how we are currently handling the MC reconstruction at all, since we aren't using the IDs to locate the geometry now in the reconstruction. It would give us the ability to speed up the reconstruction though. So I think we should proceed with the fix for that reason, viz. it will allow us to reconstruct events more quickly.

Norman



________________________________

**Comment by omar-moreno** [comment]
​OK, I was under the impression that this would cause hps-java to be
incompatible with all current MC.  Thanks for the clarification. ​

**Comment by JeremyMcCormick** [comment]
@omar-moreno @mrsolt 

Do you want me to recreate all the LCDD files based on the current tracker detector models and commit them to this branch?

This would correct the wrong IDs in the LCDD files.  It should also sync the LCDD and "Java" versions of the detector models, assuming that they are currently correct/matching.  (There was an email conversation in which we discussed that there were some changes/updates to the Java detector models that are not all reflected in the current LCDD file versions).

I don't really see a clean way to resolve the ID issue or the fact that some of the LCDD files in the repo are currently not compatible with the Java detector models other than rebuilding the LCDD files to replace the existing ones.

**Comment by JeremyMcCormick** [comment]
The mismatch between the Java and LCDD converters should be resolved by PR #192 (assuming the current versions of these classes actually match each other!).
